### PR TITLE
StackDataset batched sampling

### PR DESCRIPTION
**Pull request by stsouko**
Optimization of loading minibatches

### CODE REVIEW

**Comment by ezyang** [comment]
@ejguan will you be able to review this?

**Comment by ejguan** [comment]
> @ejguan will you be able to review this?

Sure thing. I will review it later today.

File: test/test_dataloader.py [hunk 1]
```diff
@@ -410,6 +410,35 @@ def test_getitem(self):
...
+        for i in range(4):
+            self.assertEqual(t[i], batch[i][0])
+            self.assertEqual(l[i], batch[i][1])
+
+        source = StackDataset(t=t, l=l)
+        batch = source.__getitems__([0, 1, 2, 3])
+        for i in range(4):
+            self.assertEqual(t[i], batch[i]['t'])
+            self.assertEqual(l[i], batch[i]['l'])
+
```
Comment (ejguan) [inline test/test_dataloader.py:441]: Could you pls add a test for indices out of range?

File: torch/utils/data/dataset.py [hunk 2]
```diff
@@ -251,6 +251,31 @@ def __getitem__(self, index):
             return {k: dataset[index] for k, dataset in self.datasets.items()}
         return tuple(dataset[index] for dataset in self.datasets)
 
+    def __getitems__(self, indices: list):
+        # add batched sampling support when parent datasets supports it.
+        if isinstance(self.datasets, dict):
+            dict_batch: List[T_dict] = [{} for _ in indices]
+            for k, dataset in self.datasets.items():
+                if callable(getattr(dataset, "__getitems__", None)):
+                    for data, d_sample in zip(dataset.__getitems__(indices), dict_batch):  # type: ignore[attr-defined]
```
Comment (ejguan) [inline torch/utils/data/dataset.py:260]: Could you pls add an Error if the size of result from `dataset.__getitems__(indices)` is not correct?

File: torch/utils/data/dataset.py [hunk 3]
```diff
@@ -251,6 +251,31 @@ def __getitem__(self, index):
...
+                else:
+                    for idx, d_sample in zip(indices, dict_batch):
+                        d_sample[k] = dataset[idx]
+            return dict_batch
+
+        # tuple data
+        list_batch: List[list] = [[] for _ in indices]
+        for dataset in self.datasets:
+            if callable(getattr(dataset, "__getitems__", None)):
+                for data, t_sample in zip(dataset.__getitems__(indices), list_batch):  # type: ignore[attr-defined]
```
Comment (ejguan) [inline torch/utils/data/dataset.py:271]: ditto

**Review by ejguan — COMMENTED** [review]
Overall LGTM with a few nit comments

File: torch/utils/data/dataset.py [hunk 4]
```diff
@@ -251,6 +251,39 @@ def __getitem__(self, index):
...
+    def __getitems__(self, indices: list):
+        # add batched sampling support when parent datasets supports it.
+        if isinstance(self.datasets, dict):
+            dict_batch: List[T_dict] = [{} for _ in indices]
+            for k, dataset in self.datasets.items():
+                if callable(getattr(dataset, "__getitems__", None)):
+                    items = dataset.__getitems__(indices)  # type: ignore[attr-defined]
+                    if len(items) != len(indices):
+                        raise ValueError("Nested dataset's output size mismatch."
+                                         f" Expected {len(indices)}, got {len(items)}")
```
Comment (ejguan) [inline torch/utils/data/dataset.py:263]: Can you pls add an unit test to validate the `ValueError` is raised as expectation?

**Review by ejguan — APPROVED** [review]
Thanks, overall LGTM with a comment on the test case.

**Pull request closed without merging**